pkg/storegateway/series_chunks.go

```diff
@@ -4,46 +4,148 @@ package storegateway
+		// the slice with the right size.
+		New: nil,
+	})
+
+	seriesChunksSlicePool = pool.Interface(&sync.Pool{
+		// Intentionally return nil if the pool is empty, so that the caller can preallocate
+		// the slice with the right size.
+		New: nil,
+	})
+)
+
 // seriesChunksSetIterator is the interface implemented by an iterator returning a sequence of seriesChunksSet.
 type seriesChunksSetIterator interface {
 	Next() bool
+
+	// At returns the current seriesChunksSet. The caller should (but NOT must) invoke seriesChunksSet.release()
+	// on the returned set once it's guaranteed it will not be used anymore.
 	At() seriesChunksSet
+
 	Err() error
 }
 
 // seriesChunksSet holds a set of series, each with its own chunks.
 type seriesChunksSet struct {
-	series []seriesEntry
+	series           []seriesEntry
+	seriesReleasable bool
+
+	// It gets lazy initialized (only if required).
+	seriesChunksPool *pool.SlabPool[storepb.AggrChunk]
 
 	// chunksReleaser releases the memory used to allocate series chunks.
 	chunksReleaser chunksReleaser
 }
 
+// newSeriesChunksSet creates a new seriesChunksSet. The series slice is pre-allocated with
+// the provided seriesCapacity at least. This means this function GUARANTEES the series slice
+// will have a capacity of at least seriesCapacity.
+//
+// If seriesReleasable is true, then a subsequent call release() will put the internal
+// series slices to a memory pool for reusing.
+func newSeriesChunksSet(seriesCapacity int, seriesReleasable bool) seriesChunksSet {
+	var prealloc []seriesEntry
+
+	// If it's releasable then we try to reuse a slice from the pool.
+	if seriesReleasable {
+		if reused := seriesEntrySlicePool.Get(); reused != nil {
+			prealloc = *(reused.(*[]seriesEntry))
+
+			// The capacity MUST be guaranteed. If it's smaller, then we forget it and will be
+			// reallocated.
+			if cap(prealloc) < seriesCapacity {
+				prealloc = nil
+			}
+		}
+	}
+
+	if prealloc == nil {
+		prealloc = make([]seriesEntry, 0, seriesCapacity)
+	}
+
+	return seriesChunksSet{
+		series:           prealloc,
+		seriesReleasable: seriesReleasable,
+	}
+}
+
 type chunksReleaser interface {
 	// Release the memory used to allocate series chunks.
 	Release()
 }
 
+// release the internal series and chunks slices to a memory pool, and call the chunksReleaser.Release().
+// The series and chunks slices won't be released to a memory pool if seriesChunksSet was created to be not releasable.
+//
+// This function is not idempotent. Calling it twice would introduce subtle bugs.
 func (b *seriesChunksSet) release() {
 	if b.chunksReleaser != nil {
 		b.chunksReleaser.Release()
-		b.chunksReleaser = nil
 	}
 
-	b.series = nil
+	if b.seriesReleasable {
+		// Reset series and chunk entries, before putting back to the pool.
+		for i := range b.series {
+			for c := range b.series[i].chks {
+				b.series[i].chks[c].MinTime = 0
+				b.series[i].chks[c].MaxTime = 0
+				b.series[i].chks[c].Raw = nil
```

Comment:
Are we reusing `.chks`? Because I can see them being set to `nil` below, and if so, why bother zeroing them here? Is that something that `chunkReleaser` does? In that case it would worth a comment because this code looks useless to me from this point of view.

Reply:
although if we go with my suggestion, the using SlabPool instead of BatchBytes will be more difficult because we don't want to call `.Reset()` on each byte individually

Reply:
Okay, I can see how it's reused in `b.seriesChunksPool.Release()`, my fault. I would bring this zeroing code closer to the place where it's reused. I like your suggestion too, but I'm not sure that we want to force the items being resettable.

Reply:
I don't have a big preference. I simplified the reset code in https://github.com/grafana/mimir/pull/3739/commits/59937e99b165daf8369b1cbd79964ca74e233bca and I think that's just fine. I would add further indirection.